spring-integration-file/src/main/java/org/springframework/integration/file/remote/session/ContextHolderRequestHandlerAdvice.java

```diff
@@ -0,0 +1,127 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.file.remote.session;
+
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.springframework.integration.handler.advice.AbstractRequestHandlerAdvice;
+import org.springframework.messaging.Message;
+
+/**
+ * Provides a key for the context holder.This key could for example be stored in a message header.
+ *
+ * @author Adel Haidar
+ * @since 6.1
+ */
+public class ContextHolderRequestHandlerAdvice extends AbstractRequestHandlerAdvice {
+	public Function<Message<?>, Object> keyProvider;
+
+	public Consumer<Object> contextSetHook;
+
+	public Consumer<Object> contextClearHook;
+
+	/**
+	 * Constructor for the ContextHolderRequestHandlerAdvice.
+	 *
+	 * @param keyProvider The key provider function.
+	 * @param contextSetHook The context set hook function.
+	 * @param contextClearHook The context clear hook function.
+	 */
+	public ContextHolderRequestHandlerAdvice(final Function<Message<?>, Object> keyProvider, final Consumer<Object> contextSetHook, final Consumer<Object> contextClearHook) {
+		this.keyProvider = keyProvider;
+		this.contextSetHook = contextSetHook;
+		this.contextClearHook = contextClearHook;
+	}
+
+	/**
+	 * Returns the key provider function.
+	 * @return The key provider function.
+	 *
+	 */
+	public Function<Message<?>, Object> getKeyProvider() {
+		return this.keyProvider;
+	}
+
+	/**
+	 * Sets the key provider function.
+	 * @param keyProvider the key provider.
+	 */
+	public void setKeyProvider(Function<Message<?>, Object> keyProvider) {
+		this.keyProvider = keyProvider;
+	}
+
+	/**
+	 * Returns the context set hook function.
+	 * @return The context set hook function.
+	 *
+	 */
+	public Consumer<Object> getContextSetHook() {
+		return this.contextSetHook;
+	}
+
+	/**
+	 * Sets the context set hook function.
+	 * @param contextSetHook the context set hook function.
+	 */
+	public void setContextSetHook(Consumer<Object> contextSetHook) {
+		this.contextSetHook = contextSetHook;
+	}
+
+	/**
+	 * Returns the context clear hook function.
+	 * @return The context clear hook function.
+	 */
+	public Consumer<Object> getContextClearHook() {
+		return this.contextClearHook;
+	}
+
+	/**
+	 * Sets the context clear hook function.
+	 * @param contextClearHook the context clear hook function.
+	 */
+	public void setContextClearHook(Consumer<Object> contextClearHook) {
+		this.contextClearHook = contextClearHook;
+	}
+
+	@Override
+	protected Object doInvoke(ExecutionCallback callback, Object target, Message<?> message) {
+		Object result;
+		final Object key = this.keyProvider.apply(message);
+		try {
+			setContext(key);
+			result = callback.execute();
+		}
+		catch (Exception e) {
+			result = e;
+			logger.error("Failure setting context for " + message + ": " + e.getMessage());
```

Comment:
That's not correct: the exception might be as a result of the `callback.execute()`.
I believe that we must not catch any exception over here at all and just let them to be propagated upstream as is.
Kinda not this component responsibility.